src/inmanta/server/services/compilerservice.py

```diff
@@ -1068,3 +1068,20 @@ async def environment_action_deleted(self, env: model.Environment) -> None:
     async def recalculate_queue_count_cache(self) -> None:
         async with self._queue_count_cache_lock:
             self._queue_count_cache = await data.Compile.get_total_length_of_all_compile_queues()
+
+    async def cancel_compile(self, env_id: uuid.UUID) -> None:
```

Comment:
Given that a compile may have side effects like allocation, export, ... I'm not sure if this is a good idea. As in, I honestly don't know. @wouterdb wdyt?

Apart from that, I'm afraid this will not work with the changes I made to the compiler service. For one thing, I removed the `Task` objects from `self._recompiles` (and renamed it and made it a set). But I also made it so that only a single method writes to it. I think we can extend that if we acquire the lock here, but I want to carefully consider it.

Let me know if you'd like me to make the required changes. After all, we only have this conflict because I rushed the other PR before reviewing this one.

Reply:
1. if we do this on clear only, I think we can safely do it
2. I don't know if this cancel will actually cancel the compile? (i.e. the process will keep going?)

Reply:
1. Indeed. I only call this method on clear and delete operations, because it's safe in these situations.
2. The compiler service spawns processes using `asyncio.create_subprocess_exec()`. In each situation the this code is surrounded by a try-finally block that kills the process if it's still running. So this method process guarantees that the process will be stopped by the times this method finishes execution.